Decorators and forwarding, call/apply (#204)
javascript-tutorial/it.javascript.info: 1-js/06-advanced-functions/09-call-apply-decorators/article.md

  }
};

// dovrebbe ricordare le chiamate con lo stesso argomento
worker.slow = cachingDecorator(worker.slow);
```
Precedentemente, con un singolo argomento `x` potevamo usare `cache.set(x, result)` per salvare il risultato, e `cache.get(x)` per richiamarlo. Ma ora abbiamo bisogno di memorizzare il risultato per più combinazioni di argomenti `(min,max)`, e il comando `Map` prende un solo argomento come chiave.

Sono possibili diverse soluzioni:

1. Implementare una nuova (o usarne una di terze parti) struttura simile a *map*, ma più versatile e che permetta chiavi multiple.
2. Usare *maps* annidate: `cache.set(min)` sarà un `Map` che conterrà le coppie `(max, result)`. Quindi potremo avere `result` come `cache.get(min).get(max)`.
3. Unire i due valori in uno. Nel nostro caso potemmo usare una semplice stringa `"min,max"` come chiave del `Map`. 
Per maggiore flessibilità potremmo dotare il decorator di una funzione di *hashing*, che sappia trasformare più valori in uno solo.

Per molte applicazioni pratiche, la terza soluzione è sufficiente, quindi ci atterremo ad essa.

Non abbiamo bisogno di passare solo `x`, ma anche gli altri argomenti in `func.call`.  

> Also we need to pass not just `x`, but all arguments in `func.call`. Ricordiamo che in una `function()` sono disponibili tutti i suoi argomenti tramite il *pseudo-array* `arguments`. Quindi `func.call(this, x)` può essere sostituito con `func.call(this, ...arguments)`.

Reviewer: ```suggestion
Ricordiamo che in una `function()` sono disponibili tutti i suoi argomenti tramite il *pseudo-array* `arguments`. Quindi `func.call(this, x)` può essere sostituito con `func.call(this, ...arguments)`.
```

> IL seguente è `cachingDecorator` migliorato:

Reviewer: ```suggestion
Il seguente è `cachingDecorator` migliorato:
```

worker.slow = cachingDecorator(worker.slow, hash);

alert( worker.slow(3, 5) ); // funziona
alert( "Again " + worker.slow(3, 5) ); // anche qui funziona (dalla cache)
```

Ora funziona con qualsiasi numero di argomenti (anche la funzione hash dovrebbe essere sistemata per consentire un numero qualsiasi di argomenti. Un modo interessante per farlo sarà trattato di seguito).

Ci sono du cambiamenti:

- Nella linea `(*)` viene chiamato `hash` per creare una chiave unica da `arguments`. In questo caso abbiamo usato una semplice funzione di unione che trasforma gli argomenti `(3, 5)` nella chiave `"3,5"`. Casi più complessi potrebbero richiedere approcci differenti per la funzione di *hashing*.

> - Successivamente `(**)` usa `func.call(this, ...arguments)` per passare alla funzione originale sia il contesto che tutti gli argomenti ricevuti dal *wrapper*

Reviewer: ```suggestion
- Successivamente `(**)` usa `func.call(this, ...arguments)` per passare alla funzione originale sia il contesto che tutti gli argomenti ricevuti dal *wrapper*.
```

- `apply` accetta solo *simil-array* `args`.

Quindi, se ci aspettiamo un iterabile usiamo `call`, se invece ci aspettiamo un array, usiamo` apply`.

E per oggetti che sono sia iterabili che simil-array, come un vero array, possiamo usarne uno qualsiasi, ma `apply` sarà probabilmente più veloce, perché è meglio ottimizzato nella maggior parte dei motori JavaScript.

Il passaggio di tutti gli argomenti insieme al contesto a un'altra funzione è chiamato *call forwarding* (inoltro di chiamata).

Questa è la sua forma più semplice:

```js
let wrapper = function() {
  return func.apply(this, arguments);
};
```

Quando un codice esterno chiama il `wrapper`, è indistinguibile dalla chiamata della funzione originale `func`.

## Prendere in prestito un metodo [#method-borrowing]

Ora facciamo un ulteriore piccolo miglioramento nella funzione di hashing:

```js
function hash(args) {
  return args[0] + ',' + args[1];
}
```

Per ora funziona solo su due argomenti. Sarebbe meglio se potesse unire un numero qualsiasi di `args`.

La soluzione più immediata sarebbe usare il metodo [arr.join](mdn:js/Array/join):

```js
function hash(args) {
  return args.join();
}
```

> ...Sfortunatamente non funziona, perché stiamo chiamando `hash (arguments)`, e l'oggetto `arguments` è sia iterabile che simil-array, ma non è un vero array.

Reviewer: ```suggestion
...Sfortunatamente non funziona, perché stiamo chiamando `hash(arguments)`, e l'oggetto `arguments` è sia iterabile che simil-array, ma non è un vero array.
```

hash(1, 2);
```

Il trucco è chiamato *method borrowing*.

Prendiamo (in prestito) il metodo `join` da un normale array (`[].join`) ed usiamo `[].join.call` per eseguirlo nel contesto di `arguments`.

Perché funziona?

Perché l'algoritmo interno del metodo nativo `arr.join(glue)` è molto semplice.

Preso quasi letteralmente dalla specifica:

1. Imposta `glue` come primo argomento, o, se non ci sono argomenti, una virgola `","`.
2. Imposta `result` come stringa vuota.
3. Aggiungi `this[0]` a `result`.
4. Aggiungi `glue` e `this[1]`.
5. Aggiungi `glue` e `this[2]`.
6. ...Continua fino a che `this.length` elementi sono stati "incollati".
7. Ritorna `result`.

> Quindi, tecnicamente prende `this` ed unisce `this[0]`, `this[1]` ...ecc. E' scritto intenzionalmente in moda da permette l'uso di un simil-array come `this` (non è una coincidenza se molti metodi seguono questa pratica). E' per questo motivo che funziona anche con `this=arguments`.

Reviewer: ```suggestion
Quindi, tecnicamente prende `this` ed unisce `this[0]`, `this[1]` ...ecc. E' scritto intenzionalmente in modo da permette l'uso di un simil-array come `this` (non è una coincidenza se molti metodi seguono questa pratica). E' per questo motivo che funziona anche con `this=arguments`.
```

## Decorators e proprietà di funzione

> In genere è sicuro sostituire una funzione o un metodo con uno "decorato", tranne per una piccola cosa. Se la funzione originale aveva proprietà associate, come `func.calledCount` o qualsiasi altra cosa, allora quella decorata non le fornirà. Perché quello è un *wrapper*, quindi bisogna stare attenti a come lo si usa.

Reviewer: Mi ha lasciato un attimo confuso quel "decorato", ho dovuto rileggerlo due volte. Pensavo a qualcosa di simile:

```suggestion
In genere è sicuro sostituire una funzione o un metodo con una sua versione "decorata", tranne per una piccola cosa. Se la funzione originale aveva proprietà associate, come `func.calledCount` o qualsiasi altra cosa, allora quella decorata non le fornirà. Perché quello è un *wrapper*, quindi bisogna stare attenti a come lo si usa.
```

> Es. nell'esempio sopra, se la funzione `slow` avesse avuto delle proprietà, allora `cachingDecorator (slow)` sarebbe stato un *wrapper* senza di esse.

Reviewer: ```suggestion
Es. nell'esempio sopra, se la funzione `slow` avesse avuto delle proprietà, allora `cachingDecorator(slow)` sarebbe stato un *wrapper* senza di esse.
```
